### PR TITLE
[#21577] Keycard [Wallet] - Derives next account for profile key pair

### LINKED ISSUE #21577: Keycard [Wallet] - Derives next account for profile key pair
Currently they flow of creating (deriving) new account in the wallet with keycard will end with error message "not supported", instead of that message, card should be scanned and new account (address) derived and saved similar to regular account

### LINKED ISSUE #21717: Keycard key pair has a wrong name
# Bug Report

### Reproduction

1. Install app
2. Create a user
3. Take an empty Keycard
4. Go to Profile -> Keycard
5. Go through the migration flow
6. Logout -> Log in
7. Go to Profile -> Wallet -> Key pairs

#### Actual behavior

Image

#### Expected behavior

Image

https://www.figma.com/design/QWJV5WRbroXVDsdubEDjAT/Wallet-settings-for-Mobile?node-id=47-118913&node-type=canvas&m=dev

### Additional Information

- Status version: nightly 29/11
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

### LINKED ISSUE #21749: Not a default wallet account is blocked in case we have a default account migrated on the Keycard
# Bug Report

## Problem 
[comment]: # (A problem description)

If we have 2 or more wallet accounts:

original
extra one (imported or newly created)

And then do a migration then transactions will work only for that original default migrated key pair. For extra one sign doesn't work.

**Please pay attention:** we have the same error also on Swap sometimes https://github.com/status-im/status-mobile/issues/21701

#### Actual behavior

Image

[Send_Keypair.zip](https://github.com/user-attachments/files/18006947/Send_Keypair.zip)

#### Expected behavior

Under discussion

### Reproduction

1. Install app
2. Create user by a standard flow
3. Go to Wallet -> Create new "Not a Keypair" account -> New key pair
4. Make sure you have assets on both accounts
5. Go to Profile -> Keypair and do a migration
6. Go to Wallet -> Send from "Not a Keypair" account to default one.

### Additional Information

- Status version: nightly 04/12
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

### LINKED ISSUE #21755: Sync is not available for mobile keycard accounts
# Bug Report

## Problem 
As there is no option to sync current mobile keycard wallet with desktop, users cannot continue the flows that are not available on mobile

#### Expected behavior
can sync (or use a fallback feature), see all accounts on desktop and continue to use it.

#### Actual behavior
cannot sync, as Showing QR code is blocked
Image

### Reproduction

1. Open profile > sync
2. Try to generate new QR

### Additional Information

- Status version: nightly 
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

### LINKED ISSUE #21778: Keycard - Disable or hide unsupported Keycard flows
# Problem

In Status 2.32 Keycard users won't be able to use their keycards in all possible flows. We want to provide a better user experience to those users so that they don't waste time in flows that will surely fail at some point.

## Acceptance Criteria

In case a Profile is set-up with a Keycard: 

- Request to join a community should show a drawer stopping the user from proceeding https://www.figma.com/design/YGm3igIOAcwMqUVJWCJ6f1/Keycard?node-id=4588-67142&node-type=instance&m=dev
- Show the special _unsupported feature_ drawer when the user tries to **enable** biometrics.
- Hide option to edit password.

## Out of scope

- Ignore editing shared addresses in Figma https://www.figma.com/design/YGm3igIOAcwMqUVJWCJ6f1/Keycard?node-id=4588-104535&node-type=shape_with_text&m=dev (this feature is disabled behind a feature flag, so we don't need to worry about it).
- Wallet-related flows.

## Resources

- Check PR https://github.com/status-im/status-mobile/pull/21771 because this issue can reuse some of the code related to the drawer.

### PR DESCRIPTION
**Pull request by flexsurfer**
fixes #21577
fixes #21749
fixes https://github.com/status-im/status-mobile/issues/21778
fixes https://github.com/status-im/status-mobile/issues/21755

#21717 partially fixed, only place is fixed

### CODE REVIEW

File: src/status_im/contexts/keycard/events.cljs [hunk 1]
```diff
@@ -65,6 +67,30 @@
...
+          [:keycard/connect
+           {:key-uid key-uid
+            :on-success
+            (fn []
+              (rf/dispatch
+               [:keycard/export-key
+                {:pin        pin
+                 :path       derivation-path
+                 :on-success (fn [public-key]
+                               (let [derived-account {:public-key (str "0x" public-key)
```
Comment (ilmotta) [inline src/status_im/contexts/keycard/events.cljs:82]: There's `utils.address/normalized-hex`.

File: src/status_im/contexts/keycard/events.cljs [hunk 2]
```diff
@@ -65,6 +67,30 @@
...
+   {:fx [[:dispatch
+          [:keycard/connect
+           {:key-uid key-uid
+            :on-success
+            (fn []
+              (rf/dispatch
+               [:keycard/export-key
+                {:pin        pin
+                 :path       derivation-path
+                 :on-success (fn [public-key]
```
Comment (ilmotta) [inline src/status_im/contexts/keycard/events.cljs:81]: We are two callbacks deep on this `on-success`. Would be clearer to have a separate private function for the inner `on-success` handler to improve readability.

**Review by ilmotta — APPROVED** [review]

**Review by Parveshdhull — APPROVED** [review]

**Comment by pavloburykh** [comment]
Hey @flexsurfer! Thank you for the PR. Got a question: why have we decided to replace "non supported" warning with password bottom sheet in all places where it is currently impossible to confirm action via keycard?

Now it looks more confusing. The idea behind 'not supported' warning was at least somehow show users that this functionality is unavailable under keycard. 

Based on current implementation I do not consider PR fixes https://github.com/status-im/status-mobile/issues/21749. We should either implement possibility to sign via keycard transactions from non-default key pair account OR show user 'not supported' warning as well as in all other places where we are unable to implement confirmation via keycard. Currently it looks like this:

https://github.com/user-attachments/assets/cacbe154-5af5-4cc1-b3a1-65af80bf6f6e

cc @churik @ilmotta

**Comment by flexsurfer** [comment]
> Got a question: why have we decided to replace "non supported" warning with password bottom sheet in all places where it is currently impossible to confirm action via keycard?

hey @pavloburykh good question, showing password bottom sheet is not expected and should be replaced by warning, but probably in separate issue

**Comment by pavloburykh** [comment]
> > Got a question: why have we decided to replace "non supported" warning with password bottom sheet in all places where it is currently impossible to confirm action via keycard?
> 
> hey @pavloburykh good question, showing password bottom sheet is not expected and should be replaced by warning, but probably in separate issue

Can you revert the changes in this PR, so we show 'not supported' alert in all places where keycard is not supported (as we do it in develop)? 

And regarding this issue https://github.com/status-im/status-mobile/issues/21749 - we need either remove it from current PR description (as it is not fixed) or at least add 'not supported' toast for this case.

**Comment by pavloburykh** [comment]
@flexsurfer this issue https://github.com/status-im/status-mobile/issues/21717 is not fixed as well


https://github.com/user-attachments/assets/77f32e80-f686-408a-986c-47bf284fb5ba

**Comment by flexsurfer** [comment]
> @flexsurfer this issue #21717 is not fixed as well
> 
>  telegram-cloud-document-2-5370671791205277332.mp4

thank you, yes it's kinda difficult to find all places

**Comment by flexsurfer** [comment]
> > > Got a question: why have we decided to replace "non supported" warning with password bottom sheet in all places where it is currently impossible to confirm action via keycard?
> > 
> > 
> > hey @pavloburykh good question, showing password bottom sheet is not expected and should be replaced by warning, but probably in separate issue
> 
> Can you revert the changes in this PR, so we show 'not supported' alert in all places where keycard is not supported (as we do it in develop)?
> 
> And regarding this issue #21749 - we need either remove it from current PR description (as it is not fixed) or at least add 'not supported' toast for this case.

no it's not possible to revert, we can wait https://github.com/status-im/status-mobile/issues/21763 then, i would say https://github.com/status-im/status-mobile/issues/21749 is fixed, but there are another bugs which should be fixed, but anyway we still need 21763, it's up to you , we can wait or merge it doesn't matter, both should be merged anyway before release

**Comment by pavloburykh** [comment]
> thank you, yes it's kinda difficult to find all places

You have mentioned in the description `partially fixed, only place is fixed`. Could you clarify in which place it is fixed?

**Comment by pavloburykh** [comment]
> i would say https://github.com/status-im/status-mobile/issues/21749 is fixed

just to be on the same page, could you please clarify what do we assume as a fix (expected result)? Cause I am little bit confused here. 

> but there are another bugs which should be fixed

which bugs do you mean here?

**Comment by flexsurfer** [comment]
> > thank you, yes it's kinda difficult to find all places
> 
> You have mentioned in the description `partially fixed, only place is fixed`. Could you clarify in which place it is fixed?

i mean name is not fixed, all places im aware of are fixed, but that's fine that you found some new places i'll fix it

**Comment by flexsurfer** [comment]
> > i would say #21749 is fixed
> 
> just to be on the same page, could you please clarify what do we assume as a fix (expected result)? Cause I am little bit confused here.

Not a default wallet account is NOT blocked

> > but there are another bugs which should be fixed
> 
> which bugs do you mean here?

password is shown for keycard instead of warning in some cases

**Comment by pavloburykh** [comment]
> Not a default wallet account is NOT blocked 

In terms of this [issue](https://github.com/status-im/status-mobile/issues/21749) `blocked` means empty error toast once we try to sign transaction. In this PR we show password bottom sheet instead of error toast. So it doesn't seem anything has been improved from user perspective in this PR, as user is still blocked from completing transaction and there is no clear info why he is unable to do that. In fact we can consider issue to be fixed only after new [pop up](https://github.com/status-im/status-mobile/issues/21763) is implemented and applied to this scenario. 

> no it's not possible to revert, we can wait https://github.com/status-im/status-mobile/issues/21763 then, i would say https://github.com/status-im/status-mobile/issues/21749 is fixed, but there are another bugs which should be fixed

> > which bugs do you mean here?

> password is shown for keycard instead of warning in some cases

based on this reply it is still not clear to me: in case we merge current PR into develop now, will password-bottom sheet be replaced by pop up once https://github.com/status-im/status-mobile/issues/21763 is merged or we will need to make other adjustments in order to fix the issue when password bottom sheet appears? For example, will 21763 replace password bottom sheet with pop up in [NON default wallet account scenario](https://github.com/status-im/status-mobile/issues/21749)? If adjustments will be needed, then it does not make sense to merge this PR now, better to wait for https://github.com/status-im/status-mobile/issues/21763 merged into develop, rebase current PR and make required changes. 

Sorry for a lot of questions but to many things are unclear and therefore conversations take too much time. Providing detailed description of expected result and implementation will help save time in future PRs.

**Comment by churik** [comment]
@flexsurfer any news on this PR? we're already in release testing and I don't think that it makes sense to start extensive testing of keycard without this PR. do you need help?

**Comment by flexsurfer** [comment]
sorry @pavloburykh now it should be possible to create new non keycard keypair, and generate new account for this keypair, and also sign transaction for these accounts

**Comment by pavloburykh** [comment]
> sorry @pavloburykh now it should be possible to create new non keycard keypair, and generate new account for this keypair, and also sign transaction for these accounts

@flexsurfer thanks for the fixes. Please take a look at the issue. 

### ISSUE 1 Cannot sign via Keycard creation of derived account from NON default keypair 

**Steps:** 

1. Login account with set keycard
2. Generate/import new keypair
3. Try to create additional derived account from generated/imported keyair

**Actual result:** password bottom sheet appears instead of pin bottom sheet. Cannot sign account creation. 


https://github.com/user-attachments/assets/bb75b4c4-da00-4c3a-8ede-1df4c64104f1

**Comment by pavloburykh** [comment]
### ISSUE 2 Cannot confirm swap transactions via keycard

**Steps:** 

1. Login keycard account
2. Try to perform swap transaction
3. See if you can confirm swap transaction via keycard 

**Actual result:** password bottom sheet appears instead of pin bottom sheet. Cannot confirm swap transaction via keycard


https://github.com/user-attachments/assets/35fc8746-fe51-4b4d-badb-0518b9867b21



**Expected result:** keycard support for swap has been recently implemented https://github.com/status-im/status-mobile/pull/21764 and merged into develop. Current PR is rebased so I suppose it should work.

**Comment by pavloburykh** [comment]
@flexsurfer we discussed it, but I will log it here as separate issue in order to track it

### ISSUE 3 Password bottom sheet appears instead of alert in those places where keycard action is unavailable

**Steps:** 

1. Login keycard user
2. initiate any action which requires confirmation but is not available on keycard, for example: reveal sync code, sign wallet connect event, join community etc. 
3. See if alert is shown

**Actual result:** not alert is shown. Instead of alert password bottom sheet appears. 

**Expected result:** alert should be shown. This alert will be replaced by modal in future PRs.

**Comment by pavloburykh** [comment]
@flexsurfer also, just a reminder that current [issue](https://github.com/status-im/status-mobile/pull/21753#issuecomment-2520182538) is not fixed as well. We need either to fix or to remove issue from PR description.

**Comment by flexsurfer** [comment]
thanks @pavloburykh  ISSUE 1  and "on keycard" should be fixed now

**Comment by flexsurfer** [comment]
ISSUE 2 should be fixed

**Comment by flexsurfer** [comment]
added temporary alert for join-community and setup-syncing

**Comment by flexsurfer** [comment]
wallet connect will be implemented here https://github.com/status-im/status-mobile/issues/21763

**Comment by ilmotta** [comment]
> added temporary alert for join-community and setup-syncing

@flexsurfer FYI this work is going to be implemented in https://github.com/status-im/status-mobile/issues/21778. So if you prefer you can delete the temporary alert from this PR, also to potentially avoid rebase issues in case that issue is resolved before this one.

**Comment by pavloburykh** [comment]
> > added temporary alert for join-community and setup-syncing
> 
> @flexsurfer FYI this work is going to be implemented in #21778. So if you prefer you can delete the temporary alert from this PR, also to potentially avoid rebase issues in case that issue is resolved before this one.

@ilmotta we already have this alert (see screenshot below) in develop. The problem was that in this PR at some point alert was replaced by password bottom sheet which was unexpected. @flexsurfer has fixed it and added this alert back (correct me if I am wrong @flexsurfer). So, this alert has nothing to do with new modal which is in progress in scope of https://github.com/status-im/status-mobile/issues/21778.  @ilmotta are you sure we want to merge PR with this bug (password bottom sheet instead of alert in places where keycard is unavailable)? Wouldn't it affect those devs who are currently working on implementing new modal? I am not sure which option is better and will cause less problems. I am okay with any option, just want to be sure we choose the one with will be less problematic. 

photo_2024-12-09 21 16 33

**Comment by ilmotta** [comment]
> @ilmotta we already have this alert (see screenshot below) in develop. The problem was that in this PR at some point alert was replaced by password bottom sheet which was unexpected. @flexsurfer has fixed it and added this alert back (correct me if I am wrong @flexsurfer). So, this alert has nothing to do with new modal which is in progress in scope of #21778. @ilmotta are you sure we want to merge PR with this bug (password bottom sheet instead of alert in places where keycard is unavailable)? Wouldn't it affect those devs who are currently working on implementing new modal? I am not sure which option is better and will cause less problems. I am okay with any option, just want to be sure we choose the one with will be less problematic.

@pavloburykh no objections to your recommendation 👍🏼 If @flexsurfer is mainly bringing back the same behavior as in `develop` then I agree with you let's keep things like that in this PR.

**Comment by flexsurfer** [comment]
hey @pavloburykh PR is ready for testing

**Comment by pavloburykh** [comment]
@flexsurfer thanks for the fixes. [ISSUE 3](https://github.com/status-im/status-mobile/pull/21753#issuecomment-2528185232) is fixed partially. Password drawer still appears in the following scenarios:

**1. Join community**

https://github.com/user-attachments/assets/a9b2e895-4219-434e-884a-cf06666a98fa


**2. Import keypair**

https://github.com/user-attachments/assets/8080bd8c-8719-4827-a554-a191c968573a


**3. Remove keypair**

https://github.com/user-attachments/assets/17c2960f-0048-4f2f-bc18-342897a3144c


**4. Reveal keypair QR**


https://github.com/user-attachments/assets/67076269-2ff8-43e0-b910-e2243e68bda3

**Comment by flexsurfer** [comment]
thanks @pavloburykh all should be fixed now

**Comment by pavloburykh** [comment]
> thanks @pavloburykh all should be fixed now

thanks @flexsurfer!

Now I got pin drawer on community join. Should be unavailable modal. 


https://github.com/user-attachments/assets/c8b80bcc-c2dc-4294-918d-0f23f3d1ff58

**Comment by flexsurfer** [comment]
thanks @pavloburykh  fixed

**Comment by pavloburykh** [comment]
@flexsurfer thanks for the PR and all the fixes. PR is ready for merge and cherrypick into release 2.32 branch.

**Comment by churik** [comment]
Please, let me know when it is merged and notify in [release-coordination channel ](https://discord.com/channels/1210237582470807632/1217172571917647932) to cherry-pick it, thanks!

**Pull request merged**